File: src/native/net/message/PacketCodec.cpp
#include "PacketCodec.h"

#include "BinarySerialization.h"

#include <cstring>
#include <limits>

namespace xs::net
{
namespace
{

[[nodiscard]] PacketCodecErrorCode MapSerializationError(SerializationErrorCode error_code) noexcept
{
    switch (error_code)
    {
    case SerializationErrorCode::None:
        return PacketCodecErrorCode::None;
    case SerializationErrorCode::BufferTooSmall:
        return PacketCodecErrorCode::BufferTooSmall;
    case SerializationErrorCode::LengthOverflow:
        return PacketCodecErrorCode::LengthOverflow;
    case SerializationErrorCode::InvalidArgument:
        return PacketCodecErrorCode::InvalidArgument;
    case SerializationErrorCode::InvalidBoolValue:
        return PacketCodecErrorCode::InvalidArgument;
    }

    return PacketCodecErrorCode::InvalidArgument;
}

} // namespace

std::string_view PacketCodecErrorMessage(PacketCodecErrorCode error_code) noexcept
{
    switch (error_code)
    {
    case PacketCodecErrorCode::None:
        return "Success.";
    case PacketCodecErrorCode::BufferTooSmall:
        return "Buffer does not contain enough bytes for the requested packet operation.";
    case PacketCodecErrorCode::LengthOverflow:
        return "Packet payload length exceeds the supported range.";
    case PacketCodecErrorCode::InvalidMagic:
        return "Packet header magic does not match the protocol constant.";
    case PacketCodecErrorCode::UnsupportedVersion:
        return "Packet header version is not supported.";
    case PacketCodecErrorCode::InvalidFlags:
        return "Packet header contains undefined flag bits.";
    case PacketCodecErrorCode::LengthMismatch:
        return "Packet header length does not match the provided payload bytes.";
    case PacketCodecErrorCode::InvalidArgument:
        return "Packet codec argument must not be null.";
    }

    return "Unknown packet codec error.";
}

bool IsValidPacketFlags(std::uint16_t flags) noexcept
{
    return (flags & static_cast<std::uint16_t>(~kPacketDefinedFlagMask)) == 0u;
}

PacketHeader MakePacketHeader(
    std::uint32_t msg_id,
    std::uint32_t seq,
    std::uint16_t flags,
    std::uint32_t payload_length) noexcept
{
    return PacketHeader{
        .magic = kPacketMagic,
        .version = kPacketVersion,
        .flags = flags,
        .length = payload_length,
        .msg_id = msg_id,
        .seq = seq,
    };
}

PacketCodecErrorCode ValidatePacketHeader(const PacketHeader& header) noexcept
{
    if (header.magic != kPacketMagic)
    {
        return PacketCodecErrorCode::InvalidMagic;
    }

    if (header.version != kPacketVersion)
    {
        return PacketCodecErrorCode::UnsupportedVersion;
    }

    if (!IsValidPacketFlags(header.flags))
    {
        return PacketCodecErrorCode::InvalidFlags;
    }

    return PacketCodecErrorCode::None;
}

PacketCodecErrorCode GetPacketWireSize(std::size_t payload_size, std::size_t* wire_size) noexcept
{
    if (wire_size == nullptr)
    {
        return PacketCodecErrorCode::InvalidArgument;
    }

    if (payload_size > static_cast<std::size_t>(std::numeric_limits<std::uint32_t>::max()))
    {
        return PacketCodecErrorCode::LengthOverflow;
    }

    if (payload_size > std::numeric_limits<std::size_t>::max() - kPacketHeaderSize)
    {
        return PacketCodecErrorCode::LengthOverflow;
    }

    *wire_size = kPacketHeaderSize + payload_size;
    return PacketCodecErrorCode::None;
}

PacketCodecErrorCode WritePacketHeader(const PacketHeader& header, std::span<std::byte> buffer) noexcept
{
    const PacketCodecErrorCode validation_result = ValidatePacketHeader(header);
    if (validation_result != PacketCodecErrorCode::None)
    {
        return validation_result;
    }

    if (buffer.size() < kPacketHeaderSize)
    {
        return PacketCodecErrorCode::BufferTooSmall;
    }

    BinaryWriter writer(buffer.first(kPacketHeaderSize));
    if (!writer.WriteUInt32(header.magic) ||
        !writer.WriteUInt16(header.version) ||
        !writer.WriteUInt16(header.flags) ||
        !writer.WriteUInt32(header.length) ||
        !writer.WriteUInt32(header.msg_id) ||
        !writer.WriteUInt32(header.seq))
    {
        return MapSerializationError(writer.error());
    }

    return PacketCodecErrorCode::None;
}

PacketCodecErrorCode ReadPacketHeader(std::span<const std::byte> buffer, PacketHeader* header) noexcept
{
    if (header == nullptr)
    {
        return PacketCodecErrorCode::InvalidArgument;
    }

    *header = {};
    if (buffer.size() < kPacketHeaderSize)
    {
        return PacketCodecErrorCode::BufferTooSmall;
    }

    BinaryReader reader(buffer.first(kPacketHeaderSize));
    PacketHeader parsed_header{};
    if (!reader.ReadUInt32(&parsed_header.magic) ||
        !reader.ReadUInt16(&parsed_header.version) ||
        !reader.ReadUInt16(&parsed_header.flags) ||
        !reader.ReadUInt32(&parsed_header.length) ||
        !reader.ReadUInt32(&parsed_header.msg_id) ||
        !reader.ReadUInt32(&parsed_header.seq))
    {
        return MapSerializationError(reader.error());
    }

    const PacketCodecErrorCode validation_result = ValidatePacketHeader(parsed_header);
    if (validation_result != PacketCodecErrorCode::None)
    {
        return validation_result;
    }

    *header = parsed_header;
    return PacketCodecErrorCode::None;
}

PacketCodecErrorCode EncodePacket(
    const PacketHeader& header,
    std::span<const std::byte> payload,
    std::span<std::byte> buffer) noexcept
{
    const PacketCodecErrorCode validation_result = ValidatePacketHeader(header);
    if (validation_result != PacketCodecErrorCode::None)
    {
        return validation_result;
    }

    if (payload.size() > static_cast<std::size_t>(std::numeric_limits<std::uint32_t>::max()))
    {
        return PacketCodecErrorCode::LengthOverflow;
    }

    if (header.length != static_cast<std::uint32_t>(payload.size()))
    {
        return PacketCodecErrorCode::LengthMismatch;
    }

    std::size_t wire_size = 0;
    const PacketCodecErrorCode wire_size_result = GetPacketWireSize(payload.size(), &wire_size);
    if (wire_size_result != PacketCodecErrorCode::None)
    {
        return wire_size_result;
    }

    if (buffer.size() < wire_size)
    {
        return PacketCodecErrorCode::BufferTooSmall;
    }

    const PacketCodecErrorCode write_header_result = WritePacketHeader(header, buffer.first(kPacketHeaderSize));
    if (write_header_result != PacketCodecErrorCode::None)
    {
        return write_header_result;
    }

    if (!payload.empty())
    {
        std::memcpy(buffer.data() + kPacketHeaderSize, payload.data(), payload.size());
    }

    return PacketCodecErrorCode::None;
}

PacketCodecErrorCode DecodePacket(std::span<const std::byte> buffer, PacketView* packet) noexcept
{
    if (packet == nullptr)
    {
        return PacketCodecErrorCode::InvalidArgument;
    }

    *packet = {};

    PacketHeader header{};
    const PacketCodecErrorCode read_header_result = ReadPacketHeader(buffer, &header);
    if (read_header_result != PacketCodecErrorCode::None)
    {
        return read_header_result;
    }

    std::size_t expected_wire_size = 0;
    const PacketCodecErrorCode wire_size_result =
        GetPacketWireSize(static_cast<std::size_t>(header.length), &expected_wire_size);
    if (wire_size_result != PacketCodecErrorCode::None)
    {
        return wire_size_result;
    }

    if (buffer.size() != expected_wire_size)
    {
        return PacketCodecErrorCode::LengthMismatch;
    }

    packet->header = header;
    packet->payload = buffer.subspan(kPacketHeaderSize, static_cast<std::size_t>(header.length));
    return PacketCodecErrorCode::None;
}

} // namespace xs::net

File: src/managed/Foundation/Protocol/PacketCodec.cs
using System.Buffers.Binary;

namespace XServer.Managed.Foundation.Protocol;

public enum PacketCodecErrorCode : byte
{
    None = 0,
    BufferTooSmall = 1,
    LengthOverflow = 2,
    InvalidMagic = 3,
    UnsupportedVersion = 4,
    InvalidFlags = 5,
    LengthMismatch = 6,
    InvalidArgument = 7,
}

public readonly record struct PacketView(PacketHeader Header, ReadOnlyMemory<byte> Payload);

public static class PacketCodec
{
    public static string GetErrorMessage(PacketCodecErrorCode errorCode)
    {
        return errorCode switch
        {
            PacketCodecErrorCode.None => "Success.",
            PacketCodecErrorCode.BufferTooSmall => "Buffer does not contain enough bytes for the requested packet operation.",
            PacketCodecErrorCode.LengthOverflow => "Packet payload length exceeds the supported range.",
            PacketCodecErrorCode.InvalidMagic => "Packet header magic does not match the protocol constant.",
            PacketCodecErrorCode.UnsupportedVersion => "Packet header version is not supported.",
            PacketCodecErrorCode.InvalidFlags => "Packet header contains undefined flag bits.",
            PacketCodecErrorCode.LengthMismatch => "Packet header length does not match the provided payload bytes.",
            PacketCodecErrorCode.InvalidArgument => "Packet codec argument must not be null.",
            _ => "Unknown packet codec error.",
        };
    }

    public static bool IsValidFlags(PacketFlags flags)
    {
        return ((ushort)flags & unchecked((ushort)~PacketConstants.DefinedFlagMask)) == 0;
    }

    public static PacketHeader CreateHeader(uint msgId, uint seq, PacketFlags flags, uint payloadLength)
    {
        return new PacketHeader
        {
            Magic = PacketConstants.Magic,
            Version = PacketConstants.Version,
            Flags = flags,
            Length = payloadLength,
            MsgId = msgId,
            Seq = seq,
        };
    }

    public static PacketCodecErrorCode ValidateHeader(PacketHeader header)
    {
        if (header.Magic != PacketConstants.Magic)
        {
            return PacketCodecErrorCode.InvalidMagic;
        }

        if (header.Version != PacketConstants.Version)
        {
            return PacketCodecErrorCode.UnsupportedVersion;
        }

        if (!IsValidFlags(header.Flags))
        {
            return PacketCodecErrorCode.InvalidFlags;
        }

        return PacketCodecErrorCode.None;
    }

    public static PacketCodecErrorCode GetWireSize(int payloadSize, out int wireSize)
    {
        wireSize = 0;

        if (payloadSize < 0)
        {
            return PacketCodecErrorCode.LengthOverflow;
        }

        if (payloadSize > int.MaxValue - PacketConstants.HeaderSize)
        {
            return PacketCodecErrorCode.LengthOverflow;
        }

        wireSize = PacketConstants.HeaderSize + payloadSize;
        return PacketCodecErrorCode.None;
    }

    public static PacketCodecErrorCode WriteHeader(PacketHeader header, Span<byte> buffer)
    {
        PacketCodecErrorCode validationResult = ValidateHeader(header);
        if (validationResult != PacketCodecErrorCode.None)
        {
            return validationResult;
        }

        if (buffer.Length < PacketConstants.HeaderSize)
        {
            return PacketCodecErrorCode.BufferTooSmall;
        }

        BinaryPrimitives.WriteUInt32BigEndian(buffer[..4], header.Magic);
        BinaryPrimitives.WriteUInt16BigEndian(buffer.Slice(4, 2), header.Version);
        BinaryPrimitives.WriteUInt16BigEndian(buffer.Slice(6, 2), (ushort)header.Flags);
        BinaryPrimitives.WriteUInt32BigEndian(buffer.Slice(8, 4), header.Length);
        BinaryPrimitives.WriteUInt32BigEndian(buffer.Slice(12, 4), header.MsgId);
        BinaryPrimitives.WriteUInt32BigEndian(buffer.Slice(16, 4), header.Seq);
        return PacketCodecErrorCode.None;
    }

    public static PacketCodecErrorCode ReadHeader(ReadOnlySpan<byte> buffer, out PacketHeader header)
    {
        header = default;
        if (buffer.Length < PacketConstants.HeaderSize)
        {
            return PacketCodecErrorCode.BufferTooSmall;
        }

        PacketHeader parsedHeader = new()
        {
            Magic = BinaryPrimitives.ReadUInt32BigEndian(buffer[..4]),
            Version = BinaryPrimitives.ReadUInt16BigEndian(buffer.Slice(4, 2)),
            Flags = (PacketFlags)BinaryPrimitives.ReadUInt16BigEndian(buffer.Slice(6, 2)),
            Length = BinaryPrimitives.ReadUInt32BigEndian(buffer.Slice(8, 4)),
            MsgId = BinaryPrimitives.ReadUInt32BigEndian(buffer.Slice(12, 4)),
            Seq = BinaryPrimitives.ReadUInt32BigEndian(buffer.Slice(16, 4)),
        };

        PacketCodecErrorCode validationResult = ValidateHeader(parsedHeader);
        if (validationResult != PacketCodecErrorCode.None)
        {
            return validationResult;
        }

        header = parsedHeader;
        return PacketCodecErrorCode.None;
    }

    public static PacketCodecErrorCode EncodePacket(
        PacketHeader header,
        ReadOnlySpan<byte> payload,
        Span<byte> buffer)
    {
        PacketCodecErrorCode validationResult = ValidateHeader(header);
        if (validationResult != PacketCodecErrorCode.None)
        {
            return validationResult;
        }

        if (header.Length != (uint)payload.Length)
        {
            return PacketCodecErrorCode.LengthMismatch;
        }

        PacketCodecErrorCode wireSizeResult = GetWireSize(payload.Length, out int wireSize);
        if (wireSizeResult != PacketCodecErrorCode.None)
        {
            return wireSizeResult;
        }

        if (buffer.Length < wireSize)
        {
            return PacketCodecErrorCode.BufferTooSmall;
        }

        PacketCodecErrorCode headerResult = WriteHeader(header, buffer[..PacketConstants.HeaderSize]);
        if (headerResult != PacketCodecErrorCode.None)
        {
            return headerResult;
        }

        if (!payload.IsEmpty)
        {
            payload.CopyTo(buffer[PacketConstants.HeaderSize..wireSize]);
        }

        return PacketCodecErrorCode.None;
    }

    public static PacketCodecErrorCode DecodePacket(ReadOnlyMemory<byte> buffer, out PacketView packet)
    {
        packet = default;

        PacketCodecErrorCode headerResult = ReadHeader(buffer.Span, out PacketHeader header);
        if (headerResult != PacketCodecErrorCode.None)
        {
            return headerResult;
        }

        if (header.Length > int.MaxValue)
        {
            return PacketCodecErrorCode.LengthOverflow;
        }

        PacketCodecErrorCode wireSizeResult = GetWireSize((int)header.Length, out int wireSize);
        if (wireSizeResult != PacketCodecErrorCode.None)
        {
            return wireSizeResult;
        }

        if (buffer.Length != wireSize)
        {
            return PacketCodecErrorCode.LengthMismatch;
        }

        packet = new PacketView(header, buffer.Slice(PacketConstants.HeaderSize, (int)header.Length));
        return PacketCodecErrorCode.None;
    }
}
